### DISCUSSION TITLE
Remove Cells

### DISCUSSION THREAD

**Discussion by DarkMagus** [clientIO/joint#2684] [Q&A]
### Introduction

I have problem deleting Elements from a graph. The following code is trying to delete 3 sibling elements  linked to a parent element, but the elements remain in the graph after deletion. Am I messing something? Thanks lot.
```
paper.freeze();
graph.removeCells([graph.getCell('ft-1227795'), graph.getCell('ft-1227796'), graph.getCell('ft-1227797')]);
paper.unfreeze();
const cell = graph.getCell('ft-1227796');  // This cell should have been removed, so graph.getCell('ft-1227796') is expected to return null
console.log('cell', cell);   // However, the cell appears to still be present
 ```

The output of console.log shows that the graph field of the deleted cell is null. How come that graph still references the cell?
![image](https://github.com/clientIO/joint/assets/4059723/bb6c0976-2e4e-48c4-adf7-5f359d5f75d8)

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS+

**Comment by kumilingus** [comment]
What are the steps to reproduce this? It works as expected when I try.

**Reply by DarkMagus** [reply]
@kumilingus Thanks lot for replying quickly

It is part of a code base that unfortunately cannot share without the approval of my client which may take weeks. The objective is simply to remove a whole subgraph. I found that graph.reset fixes my problem:
```
const newCells = graph.getCells().filter((cell) => !faultTreeBranch.includes(cell));
paper.freeze();
graph.resetCells(newCells)
paper.unfreeze();
```

Although, I have no idea why the graph.removeCells does not work in this case

**Reply by kumilingus** [reply]
But this will re-render all the remaining views. 

You probably manipulated the cells in some way before removing them. Without the specific steps, there is not much we can do.

**Reply by DarkMagus** [reply]
I agree that the above code will re-render all the remaining views but unfortunately I have no better solution: graph.removeCells just does not work for me. 
You said: "*You probably manipulated the cells in some way before removing them*". What are the kind of cell manipulations that will stop a graph from removing a cell? JointJS documentation does not tell anything of such kind.
The JointJS+ documentation says simply that **graph.removeCells(cells[,opt]) removes the given cells from the graph** without much fuss.

**Reply by kumilingus** [reply]
It could be that the cell was part of a different graph before. I am not sure. It could be also a bug on our site.
You can try to reproduce the issue from scratch.

**Reply by DarkMagus** [reply]
@kumilingus - 
You said: ***It could be that the cell was part of a different graph before.***

Can this be the issue? Let me explain to you what I am doing.
I have a fault tree (see below figure) that looks as follow:
1- a dendogram tree that goes from left to right
2- each node of dendogram may contain a fault tree that when expanded will show a fault tree going from top to bottom (see second figure)
3- if a top-bottom fault tree is already open, this top-bottom fault tree should be wiped out before expanding another new fault tree


The dendogram part looks as follow:
![image](https://github.com/clientIO/joint/assets/4059723/c73628f2-c6a3-48c6-ac18-6f81cfa52457)


When a node is expanded, the graph will look like the following:
![image](https://github.com/clientIO/joint/assets/4059723/9added67-14b3-4b25-90aa-be6879eef11c)




When I expand a dendogram node, I create a new temporary graph, populate it with nodes, do the top-bottom layout,   and add the new cells to the dendogram as follow:
dendogram.addCells(faulttree.getCells());


could this be the issue because the temporary faulttree graph and dendogram share cells?  

if that is the case, does JointJS+ offer any function that allows users to addCells from another graph and take the whole ownership of the cells?

**Reply by kumilingus** [reply]
That will be the problem. A cell cannot be member of more than one graph. If you add a cell to a temporary graph and then to the dendogram graph, the cell will not become a member of dendogram (its `graph` reference points to the temporary graph).
- Removing the cells from the temporary graph first should help.
- Or using the `dry: true` flag when adding the cells to the temporary graph
    ```ts
    tmpGraph.addCells(cells, { dry: true });
    ```

Note that we use the flag in [DirectedGraph](https://github.com/clientIO/joint/blob/master/packages/joint-layout-directed-graph/DirectedGraph.mjs#L106) too.

**Reply by DarkMagus** [reply]
@kumilingus Thanks for your quick answer. 
I did not try yet the second option with dry: true.  However, I tried before option 1 by trying to clear the temporary graph after I copied its cells to the dendogram graph but I had the following issues:  

1. tmpGraph.clear() crushed my program
2. tmpGraph.resetCells([]) produces strange results on the dendogram graph where some linkes (not all) disappeared from the dendogram

**Reply by kumilingus** [reply]
Please try the other one then.

> tmpGraph.clear() crushed my program

What do you mean?

> tmpGraph.resetCells([]) produces strange results on the dendogram graph where some linkes (not all) disappeared from the dendogram

This is a bug. `cell.collection` is undefined. `cell.graph` stays the same.

**Reply by DarkMagus** [reply]
@kumilingus
The  ```dry: true``` flag solves the problem. Now, ```graph.removeCells``` works as expected. Thanks lot for your help

Regarding  ```tmpGraph.clear()``` which crushed my program, I do not remember the exact error message but it has to do with LinkView.
If you like I can retry to produce the issue and send you the exact error message.

Thanks again for your help and patience!

**Reply by DarkMagus** [reply]
@kumilingus 

As promised, here is the error message I got when calling ```tmpGraph.clear```:
![image](https://github.com/clientIO/joint/assets/4059723/5877bf1c-b9aa-461e-b4ca-f3b9c831970a)

**Reply by kumilingus** [reply]
| `tmpGraph.resetCells([])` produces strange results on the dendogram graph where some linkes (not all) disappeared from the dendogram

Fixed here: https://github.com/clientIO/joint/pull/2692

> Error: LinkView: invalid target cell

It means there is no cell with **target id** in the graph. The  `graph.clear()` seems to remove the references as expected.

**Comment by DarkMagus** [comment]
@kumilingus 
After calling dendrogram.removeCells(cells), the dendogram's node containing the deleted fault tree does not show the right collapse button: it should show + button but it still showing - button despite the fact I have set its status to ```node.set('collapsed', true)```.


![image](https://github.com/clientIO/joint/assets/4059723/0fd827ff-09ce-4445-80ce-1978a261be41)

 How can I force JointJS+ to redraw the node in order to show the right collapse sign?
The only way I found to force it to show the right collapse button is using this little trick:
```graph.translate(0.0001, 0.0001);```

is there a JointJS+ better way of achieving the trick?

Thanks

**Reply by kumilingus** [reply]
And how's your collapse button implemented?

**Reply by DarkMagus** [reply]
@kumilingus 
It is a simple code that I took from one of JointJS+ app. It works well when the whole graph is drawn like it was the case when I was resetting dendogram nodes from the tempGraph because this forces the redraw of nodes, but with removeCells is no longer the case as you mentioned in one of our exchanges :

```
export const ExpandButton = joint.elementTools.Button.extend({
    options: {
        x: 'calc(w / 2)',
        y: 'calc(h + 10)',
        action: (evt: any, view: any) => {
            view.paper.trigger('element:expand', view, evt);
        },
        markup: [
            {
                tagName: 'rect',
                selector: 'button',
                attributes: {
                    'fill': '#cad8e3',
                    'x': -8,
                    'y': -8,
                    'width': 16,
                    'height': 16,
                    'cursor': 'pointer',
                }
            },
            {
                tagName: 'path',
                selector: 'icon',
                attributes: {
                    'fill': 'none',
                    'stroke': '#131e29',
                    'stroke-width': 2,
                    'pointer-events': 'none',
                }
            }
        ]
    },
    update() {
        // @ts-ignore
        joint.elementTools.Button.prototype.update.call(this, arguments);
        this.childNodes.icon.setAttribute('d', this.getIconPath());
    },
    getIconPath() {    
        if (this.relatedView.model.get('collapsed')) {
            return 'M -4 0 4 0 M 0 -4 0 4';
        } else {
            return 'M -4 0 4 0';
        }
    }
    
});
```


```
export function addTools(paper: joint.dia.Paper, elements: any[]) {
    const toolName = 'expand-tools';
    elements.forEach((element) => {
        const view = element.findView(paper);
        if (view.hasTools(toolName)) return;
        const toolsView = new joint.dia.ToolsView({
            name: toolName,
            tools: [
                new ExpandButton()
            ]
        });
        view.addTools(toolsView);
    })
}
```



in the 'element:expand' event trigger, I basically set collapsed attribute to either true or false

**Reply by kumilingus** [reply]
You can
- call `expandButton.update()` directly
- request an update with `elementView.requestUpdate(joint.dia.CellView.Flags.TOOLS);`
- implement the button as a [custom highlighter](https://docs.jointjs.com/api/dia/HighlighterView/#update_attributes) and use [UPDATE_ATTRIBUTES](https://docs.jointjs.com/api/dia/HighlighterView/#update_attributes) as shown in the [dwdm](https://github.com/clientIO/joint/blob/master/examples/dwdm/src/shapes/highlighters.ts#L48-L77) example.

**Reply by DarkMagus** [reply]
@kumilingus 
I tried the following:

1. Calling ```expandButton.update()``` directly - It works correctly
2. Requesting an update with ```elementView.requestUpdate(joint.dia.CellView.Flags.TOOLS);```  - It does not pass the compilation phase, I got the following error:  ```Property 'Flags' does not exist on type 'typeof CellView'.```
3. I did not try the third option as I do not need a highlighter to do what I want.

Although option 1 is working correctly, I would like to know why the second option is not working (just to advance my JointJS+ knowledge)

Thanks

**Reply by kumilingus** [reply]
My bad. It supposed to be like this:
```ts
elementView.requestUpdate(elementView.getFlag(joint.dia.ElementView.Flags.TOOLS));
```

**Reply by DarkMagus** [reply]
@kumilingus 
I will try it and let you know - Thanks lot

**Reply by DarkMagus** [reply]
@kumilingus 

Another question - I looked everywhere in JointJS+ documentation and could not find any entry for ```requestUpdate```  or ```getFlag```

Am I looking into the wrong documentation page?

Thanks

**Reply by kumilingus** [reply]
This topic is not covered in the documentation yet. We need to port the article from the [old docs](https://resources.jointjs.com/docs/jointjs/v4.0/joint.html#dia.CellView.custom) and extend it.
